src/librustc/session/mod.rs

```diff
@@ -1210,7 +1221,34 @@ pub fn build_session_(
         has_global_allocator: Once::new(),
         has_panic_handler: Once::new(),
         driver_lint_caps: Default::default(),
+        dir_files_map: DirFilesMap::default(),
+    };
+
+    // The session's `dir_files_map` is currently empty; time to populate it.
+    // We traverse over all the directories that can be searched by the two
+    // kinds of `FileSearch` ("target" and "host"), and get the names (as
+    // `PathBuf`s) of all the files within each directory.
+    //
+    // There is some overlap between the directories traversed by the two
+    // `for_each_lib_search_path` calls, but `add_dir_to_map` avoids calling
+    // `read_dir` on any directory more than once.
+    let mut dir_files_map = DirFilesMap::default();
+    let mut add_dir_to_map = |path: &Path, _kind| {
+        dir_files_map.entry(path.to_path_buf()).or_insert_with(|| {
+            match std::fs::read_dir(path) {
+                Ok(files) => {
+                    files.filter_map(|p| {
+                        p.ok().map(|s| s.path())
+                    })
+                    .collect::<Vec<_>>()
+                }
+                Err(..) => vec![],
+            }
+        });
     };
+    sess.target_filesearch(PathKind::All).for_each_lib_search_path(&mut add_dir_to_map);
+    sess.host_filesearch(PathKind::All).for_each_lib_search_path(&mut add_dir_to_map);
```

Comment:
I think maybe what `{target,host}_filesearch` compute on the fly should also be cached - that would allow to include the `add_dir_to_map` values alongside those "filesearch" directories without any hashmaps.

Reply:
`{target,host}_filesearch` don't compute much. They effectively do a `filter` of the session's short list of search paths -- excluding paths with non-matching kinds -- and then compute one additional path (`tlib_path`) from the sysroot and triple. Caching won't win anything here.

- `target_filesearch` and `host_filesearch` have overlap in the directories they consider. In fact, on the examples I looked at, they were identical, presumably because I wasn't cross-compiling.

`{target,host}_filesearch` take a `kind` parameter, which acts as a filter on which search paths they consider.